[team6_pair1] 자동차 경주게임 2단계 구현 완료하였습니다.

kwangilcho opened the pull request:
1. 경주하는 자동차에 이름을 추가하는 기능을 구현했습니다.
2. 우승한 자동차를 출력하는 기능을 구현하였습니다.

Commit:
자동차 경주게임 2단계 구현 완료하였습니다.

1. 경주하는 자동차에 이름을 추가하는 기능을 구현했습니다.
2. 우승한 자동차를 출력하는 기능을 구현하였습니다.

javajigi commented on src/main/java/racingcar/Car.java:
```diff
@@ -1,10 +1,15 @@
-public class Car {
+public class Car implements Comparable<Car> {
+    private String name;
     private int position = 0;
```
클래스의 구현 순서는 다음과 같다. 다음 원칙에 따라 구현한다.

```
class A {
    상수(static final) 또는 클래스 변수
    인스턴스 변수
    생성자
    메소드
}
```

javajigi commented on src/main/java/racingcar/Car.java:
```diff
@@ -1,10 +1,15 @@
     static final int MOVE_THRESHOLD = 4;
     static final String MOVE_MARKER = "-";
 
+    public Car(String name) {
```
name이 null인 경우는 체크하지 않나요?

javajigi commented on src/main/java/racingcar/RacingCarGame.java:
```diff
@@ -28,23 +25,47 @@ public void run(int numTrial) {
     }
 
+    public void registerIfWinner(List<String> winners, Car car) {
+        if (cars.get(0).getPosition() == car.getPosition())
```
상태를 가지는 객체의 데이터를 꺼내려(get)하지 말고 객체에 메시지를 보내라는 원칙에 따라 구현한다.

javajigi commented on src/main/java/racingcar/RacingCarGame.java:
```diff
@@ -28,23 +25,47 @@ public void run(int numTrial) {
+        }
+    }
+
+    public List<String> getWinners() {
```
클래스 메소드로 구현하는 경우 테스트는 어떻게 달라질까요?

javajigi commented on src/main/java/racingcar/RacingCarGame.java:
```diff
@@ -1,17 +1,14 @@
-import java.util.ArrayList;
+import java.util.*;
 
 public class RacingCarGame {
```
RacingGame이 사용자에게 값을 입력받고, 결과를 출력하고, 로직을 구현하고..
너무 많은 일을 하고 있다. 객체가 한가지 역할만 가지도록 분리한다.
MVC 패턴 기반으로 구현하는 연습을 해본다.

javajigi commented on src/test/java/racingcar/RacingCarGameTest.java:
```diff
@@ -11,6 +11,7 @@
-        game = new RacingCarGame(5);
+        String[] names = {"호눅스", "크롱", "포비"};
+        game = new RacingCarGame(names);
     }
```
우승자를 구하는 부분이 복잡한 로직으로 보여지는데요.
단위 테스트가 없어 아쉽네요. 단위 테스트 구현을 한번 도전해 보면 어떨까요?

javajigi reviewed (changes requested):
단위 테스트가 힘들지만 한번 도전해 보면 어떨까요?

Commit:
문자열 덧셈 계산기를 구현하였습니다.

1. 기본 구분자를 이용한 계산을 구현하였습니다.
2. 커스텀 구분자를 이용한 계산을 구현하였습니다.
3. 다양한 예외를 처리하였습니다.
    - 빈 문자열, 음수 값, 숫자 이외의 값
4. 테스트 코드 작성 및 유닛 테스트를 수행하였습니다.

Commit:
자동차 경주 게임 STEP2 피드백 적용했습니다.

1. MVC 패턴을 적용했습니다.
2. dto를 활용했습니다.
3. 유닛 테스트 코드를 작성했습니다.

TODO:
    1. RacingCarGame의 run 메소드를 테스팅해야합니다.
    2. RacingCarGame의 getWinner 메소드를 테스팅해야합니다.

javajigi reviewed (changes requested):
MVC 패턴 기반으로 구현 잘 했네요. pr이 늦었지만 그 만큼 깔끔하게 구현 잘 했네요. 💯
